### PR TITLE
Add link to latest test in ObsRsync Project list

### PR DESCRIPTION
**Pull request by andrii-suse**

### CODE REVIEW

File: lib/OpenQA/WebAPI/Plugin/ObsRsync.pm [hunk 1]
```diff
@@ -261,6 +262,13 @@ sub _get_last_test_id {
     return $res;
 }
 
+sub _get_test_result {
+    my ($c, $id) = @_;
+    my $job;
+    eval { $job = $c->schema->resultset("Jobs")->single({id => $id}); };
```
Comment (Martchus) [inline lib/OpenQA/WebAPI/Plugin/ObsRsync.pm:268]: What kind of errors are you hiding here? We still don't wrap database queries in general within eval blocks. This query is not supposed to fail unless the entire database or database connection is on fire and in these cases it is ok to return a 500 error, print the error to the log and stop further processing. But I'd also be interested what we others think about this.

Comment (andrii-suse) [reply lib/OpenQA/WebAPI/Plugin/ObsRsync.pm:268]: These functions build UI. It is important that UI is usable in extreme edge cases. If database is on fire - it will be visible in other places and UI doesn't need dozens errors popping up anywhere, neither in logs. It is OK that some elements in UI will not show some information, as long as UI remains usable overall.

Comment (okurz) [reply lib/OpenQA/WebAPI/Plugin/ObsRsync.pm:268]: both of you have a point :)

Comment (andrii-suse) [reply lib/OpenQA/WebAPI/Plugin/ObsRsync.pm:274]: I do understand reasons for "never hide errors" approach. But the main point of the Plugin is to help to make decisions regarding testing and releasing OBS deliverables. It is important to show information which can help to make those decisions in a consistent and friendly manner. Should some optional information be unavailable for some reasons - it is not big deal - the users should be able to make right move without it, or retrieve it from other sources. But they are definitely not interested to see occasional DB errors, especially when everything is on fire. 
The same for API request - it should return job_id of the test and may try to retrieve result of that job as well. It is caller's decision what to do with test_id when test result cannot be accessed for some reasons. But they don't want to get 500 response and wonder what's the problem, especially when there is no critical problem for what they requested and some functionality can try to remain robust even when database is on fire or something is experiencing temporary glitch or DB simply has to restart for some reasons, etc
I understand that reviewers would be happier if I created two versions of the same function - one throws errors, the other hides them when needed (i.e. always). But I don't really want to go this way and do want app logs be more friendly.
In other words - I do believe that software is written for users, not for developers. If at some point somebody is interested in occasional DB problems when building UI - they may have an idea where and what they what to store on those errors. But "stop building UI if DB has problem" is definitely not what users need in this particular case.

Comment (Martchus) [reply lib/OpenQA/WebAPI/Plugin/ObsRsync.pm:280]: > But they are definitely not interested to see occasional DB errors, especially when everything is on fire.

They will not see them unless they look at the server log. Unlike in your development setup the Mojolicious framework is run in production mode on our production instances and users will just get a 500 response without any further details (which could even leak sensible information). So the eval hides information from *developers/admins*, not from the user. That the eval hides errors which might be useful for debugging is actually the main problem I have with it.

> But they don't want to get 500 response

I get your point that it is better for the user to see at least some information instead of just getting a 500 response. However, the error should then still be visible on the server log and not be completely ignored.

I'm also wondering whether this `eval` block would really help to prevent a 500 response in case the overall database access is broken. There's often a lot going on when serving a request and the database is likely used elsewhere anyways without `eval`. So likely that effort is just pointless.

Comment (andrii-suse) [reply lib/OpenQA/WebAPI/Plugin/ObsRsync.pm:280]: Exactly - my point is that I cannot imagine situation when troubleshooting is needed of that line of code and thus I don't see any benefit of having that info in logs. And I am against the approach of flooding logs with all info you get 'just in case' and logging is needed only when you understand how to use it. And from the other side - if some DB problem happens - then flood of messages 'There is a DB problem' from every possible place will not help troubleshooting at all. 
And - yes, DB is mostly unused inside UI of the Plugin and it brings mostly supportive information, which is ok to hide occasionally.

Comment (Martchus) [reply lib/OpenQA/WebAPI/Plugin/ObsRsync.pm:280]: Admittedly, when the database access is completely broken we don't benefit from the additional log message. But you also have to see that the eval is likely pointless because the route will likely not being rendered anyways when the database access is completely broken.

I'd also like to note that the `eval` hides *all kinds* of programming errors which are potentially within our code an need fixing. Let's assume one calls the function with an `$id` which leads to an DBIx exception (e.g. `$id` is some hash which DBIx tries to interpret as advanced search parameter). That would be clearly a bug within our code that should be fixed. The `eval` prevents us from finding it.

Comment (kraih) [reply lib/OpenQA/WebAPI/Plugin/ObsRsync.pm:280]: Using `eval {...}` without `$@` handling in cases like this is not defensive programming. It's just hiding potential bugs. If this API is so unstable that it would flood the logs with error messages, then it needs to be fixed.

Comment (andrii-suse) [reply lib/OpenQA/WebAPI/Plugin/ObsRsync.pm:280]: > If this API is so unstable that it would flood the logs with error messages

The question is about what behavior you want to have when 'DB is having hard time' and your comment is missing the point completely.

Comment (andrii-suse) [reply lib/OpenQA/WebAPI/Plugin/ObsRsync.pm:280]: > But you also have to see that the eval is likely pointless because the route will likely not being rendered anyways when the database access is completely broken

I don't see any reason why routes which don't need DB access at all would be broken when DB is down.

> The eval prevents us from finding it

I believe that error will be in postgres log, I don't see much benefit from having it in app log as well. If this is kind of red flag and you guys are sure you clearly want to see thousands useless errors when real troubleshooting is required - I definitely will follow your recommendations.

File: lib/OpenQA/WebAPI/Plugin/ObsRsync.pm [hunk 2]
```diff
@@ -261,6 +276,25 @@ sub _get_last_test_id {
     return $res;
 }
 
+sub _get_test_result {
+    my ($c, $id) = @_;
+    my $job;
+    eval { $job = $c->schema->resultset("Jobs")->single({id => $id}); 1; }
+      or log_error("Error while trying to retrieve job in _get_test_result: $_");
```
Comment (kraih) [inline lib/OpenQA/WebAPI/Plugin/ObsRsync.pm:283]: That looks like it should be using `$@`.

File: lib/OpenQA/WebAPI/Plugin/ObsRsync.pm [hunk 3]
```diff
@@ -274,7 +308,9 @@ sub _get_run_last_info {
 
     my $linkpath = Mojo::File->new($home, $project, $batch, '.run_last');
     my $folder;
-    eval { $folder = readlink($linkpath) };
+    eval { $folder = readlink($linkpath); 1; }
+      or log_error("Cannot read symbolic link in _get_run_last_info: $_");
```
Comment (kraih) [inline lib/OpenQA/WebAPI/Plugin/ObsRsync.pm:312]: Probably should use `$@` too.

File: lib/OpenQA/WebAPI/Plugin/ObsRsync/Controller/Folders.pm [hunk 4]
```diff
@@ -197,26 +205,23 @@ sub forget_run_last {
...
-sub latest_test {
-    my $self  = shift;
-    my $alias = $self->param('alias');
-    my $full  = $self->param('full');
-    my $id    = $self->obs_rsync->get_last_test_id($alias);
+sub test_result {
+    my $self    = shift;
+    my $alias   = $self->param('alias');
+    my $version = $self->param('version');
+    my $full    = $self->param('full');
```
Comment (kraih) [inline lib/OpenQA/WebAPI/Plugin/ObsRsync/Controller/Folders.pm:212]: Form validation would make sense here.

File: lib/OpenQA/WebAPI/Plugin/ObsRsync.pm (hunk 3 above)
Comment (kraih) [reply lib/OpenQA/WebAPI/Plugin/ObsRsync.pm:312]: Actually, `readlink` usually does not raise errors but return an undef value (with `$!`). So this code is wrong.

Comment (andrii-suse) [reply lib/OpenQA/WebAPI/Plugin/ObsRsync.pm:312]: Exactly - usually. And if it 'sometimes' throws an exception - it is enough reason to wrap it with eval, otherwise UI will not load at all in those cases instead of showing what it can.

File: lib/OpenQA/WebAPI/Plugin/ObsRsync.pm (hunk 1 above)
Comment (Martchus) [reply lib/OpenQA/WebAPI/Plugin/ObsRsync.pm]: Ok, maybe this even works when the whole database isn't available. I haven't traced where this is called. I'm just saying that most routes will be unavailable in that case, e.g. the overall UI would most certainly not load anymore because the database is required for the navbar.

> I believe that error will be in postgres log

If the potential programming/setup error produces a PostgreSQL query. But maybe DBIx won't even produce a query because it fails before.

Comment (Martchus) [reply lib/OpenQA/WebAPI/Plugin/ObsRsync.pm]: > you guys are sure you clearly want to see thousands useless errors when real troubleshooting is required

It is true that when the DB is down we get lots of errors in the log (one per affected request and almost all requests are affected). So this doesn't make it much worse. However, if the *wrapped code* is really wrong (e.g. bug in DBIx, wrong `$id` being passed) we will not be able to find out about it because the error is hidden.

Comment (andrii-suse) [reply lib/OpenQA/WebAPI/Plugin/ObsRsync.pm]: > But maybe DBIx won't even produce a query because it fails before.

> However, if the wrapped code is really wrong

Then not only the corresponding tests will fail, but functionality wouldn't be available. And if you don't care about functionality - then practically there would be no difference at all, at least no difference which can justifiy these discussions.

Comment (Martchus) [reply lib/OpenQA/WebAPI/Plugin/ObsRsync.pm]: Then add a comment like `# we don't care at all if the information from the job is not accessible` because usually it makes a difference for us when certain functionality does not work. But when you don't care about this functionality, why do you put effort in implementing it in the first place? And since the code base is large enough maybe we could get rid of unneeded code then?

Comment (andrii-suse) [reply lib/OpenQA/WebAPI/Plugin/ObsRsync.pm]: It is you who don't care about functionality - me (and hopefully users of the UI) will complain if something doesn't work.

> Then add a comment like

It is not exactly `we don't care if information is not available`, more like `it is not critical if information is not available.` 
And doesn't you try to comment exactly what expression does? I think it is obvious that If something is wrapped in `eval`, then failure in that code is not critical and shouldn't break execution?

Comment (kraih) [reply lib/OpenQA/WebAPI/Plugin/ObsRsync.pm]: Good comments generally answer "Why does it do this?", and not "What does it do?".

Comment (kraih) [reply lib/OpenQA/WebAPI/Plugin/ObsRsync.pm]: I don't think `eval` (or `try` for that matter) says anything about how critical an exception is. It just says that you want to handle exceptions in that block of code yourself, for a reason you probably will have to explain in more detail.

Comment (kraih) [reply lib/OpenQA/WebAPI/Plugin/ObsRsync.pm]: But let us take a step back. What exceptions specifically do you want to ignore here? In what cases does `->single({id => $id})` throw? Can you maybe replace that with a method/SQL query that handles the case more gracefully?

Comment (andrii-suse) [reply lib/OpenQA/WebAPI/Plugin/ObsRsync.pm]: I did write few times already that I want to make sure UI remains as robust as possible in every edge case scenario. If some property of unimportant UI control doesn't load - it is not big deal and most of functionality will remain useful even when some parts of system experience any kind of issues - permanent or transient.

Comment (kraih) [reply lib/OpenQA/WebAPI/Plugin/ObsRsync.pm]: Then using `eval` here is not justified.

Comment (andrii-suse) [reply lib/OpenQA/WebAPI/Plugin/ObsRsync.pm]: Care to explain why? DBIx is full of throw_exception(). What level of attitude gives you permission to decide that UI should be completely unavailable and all functionality must wait next hotpatch e.g. if somebody renames the table and forgets to adjust all occurrences in source code? Or you do pretend that bugs don't happen and you can control what is happening in all places and what will be happening in future?

Comment (kraih) [reply lib/OpenQA/WebAPI/Plugin/ObsRsync.pm]: Since my opinion doesn't appear to be enough here i've asked the author of `DBIx::Class` for a second opinion:
```
22:41 <kraih> mst: do you believe it is a good idea to wrap every "...->resultset(...)->single({id => 123})" call in an eval as a measure of defensive programming?
22:42 — mst throws up in his mouth
22:42 <mst> mstcat does NOT approve of this.
22:42 <kraih> :D
22:43 <mst> DBIC will return either a single row or undef
22:43 <kraih> co-worker won't believe me, so i was hoping for something quote worthy from you
22:44 <mst> the only reasons that it would throw an exception are (1) your schema is broken, in which case fix your CI (2) not only is your $dbh disconnected but DBIC fails to reconnect as well
22:45 <mst> at which point bailing out the request is almost certainly the Correct thing to do anyway so you don't basically DoS yourself and/or your database
22:45 <mst> realistically, wrapping that in an eval is like wrapping twelve layers of clingfilm around a box that you're shipping a live cat in. no matter how good an idea it seems like, what arrives at the other end is going to make schroedinger cry.
```
I hope that settles the issue once and for all.

Comment (andrii-suse) [reply lib/OpenQA/WebAPI/Plugin/ObsRsync.pm]: Even your question has no relation to the discussion. It is not "every call" it is "the only call in module". And even his reply that there are cases when that code can throw errors didn't ring the bell for you, because it is one of cases I wanted to prevent from. E.g. I want the form to load even when the table is missing for some reasons or read locked. Is it enough reason?

File: lib/OpenQA/WebAPI/Plugin/ObsRsync.pm [hunk 5]
```diff
@@ -274,7 +308,13 @@ sub _get_run_last_info {
 
     my $linkpath = Mojo::File->new($home, $project, $batch, '.run_last');
     my $folder;
-    eval { $folder = readlink($linkpath) };
+    eval {
+        $folder = readlink($linkpath);
```
Comment (kraih) [inline lib/OpenQA/WebAPI/Plugin/ObsRsync.pm:312]: In which case do you believe this `readlink` call will throw an exception?

Comment (andrii-suse) [reply lib/OpenQA/WebAPI/Plugin/ObsRsync.pm:312]: Let's assume I want to try the plugin on some exotic FS or let's in Windows. Other functionality will be still available, but because of this line - I couldn't try it at all. It is not question about thinking all possible scenarios - it is you attitude to what you want to have.

Comment (kraih) [reply lib/OpenQA/WebAPI/Plugin/ObsRsync.pm:312]: Windows support is completely out of the question, we rely much too strongly on features Windows Perl could never support (i know that because many of those features are used in Mojolicious and Minion). And i would very much appreciate it if you would refrain from ad-hominem attacks in the future.

File: lib/OpenQA/WebAPI/Plugin/ObsRsync.pm (hunk 1 above)
Comment (kraih) [reply lib/OpenQA/WebAPI/Plugin/ObsRsync.pm]: To state the problem more clearly. The `->single` call above will only throw an exception in very serious situations, such as when the database schema is corrupted. Situations that always benefit from quick intervention by the maintainers.

So by making this a non-fatal error, several problems will arise, a) the problem will be less likely to be noticed in CI than a fatal error (log messages are easy to overlook while exceptions stop execution of the code), b) there is a big risk of further data corruption because execution did not stop here, c) the user will think everything is ok and not open a ticket (slowing down response), d) the standard framework exception handling will be disabled (which might have extra notification features enabled).

And that's just what i could think of at the moment, i'm sure there is more.

Comment (andrii-suse) [reply lib/OpenQA/WebAPI/Plugin/ObsRsync.pm]: If the problem is really fatal - it will be noticed in other places. It is not fatal for this module, permanent or transient problems with DB are not enough reason to disable sync functionality. E.g. you wouldn't log out existing sessions "to stop data corruption" if login functionality doesn't work for some reasons. CI does check whether code inside eval functions correctly. So, CI failure will or will not happen when and only when functionality inside block is covered by tests. I don't see how CI can be reason for catching or not catching schema problems here, at least I do not see strong arguments here. But another scenario may happen : the tests results are ignored or tests are disabled by mistake or intentionally. Then all rsync functionality will be unavailable until patch because of non-critical issue. (Unless developers care to distinguish what is important and what is not).

File: lib/OpenQA/WebAPI/Plugin/ObsRsync.pm (hunk 5 above)
Comment (andrii-suse) [reply lib/OpenQA/WebAPI/Plugin/ObsRsync.pm:312]: It is not about support - not sure where it came from. It is about tolerance of edge case scenarios. It can even be some experimental custom build of perl or drivers or hardware etc - of course it is not supported, but still it can be useful in some scenarios. No need to pretend every possible edge case is known in advance. So why wouldn't just protect execution flow if particular failure is possible and can be tolerated (without regard how likely is the failure - "it is possible" and "it can be tolerated" is enough).

Comment (kraih) [reply lib/OpenQA/WebAPI/Plugin/ObsRsync.pm:312]: Perhaps you should have taken a look at the [readlink](https://metacpan.org/pod/distribution/perl/pod/perlfunc.pod#readlink1) documentation first. It says exactly when it will throw an exception. The [reasons](https://metacpan.org/pod/distribution/perl/pod/perlport.pod#readlink) are very very narrow, and exclusively portability. The only justification for an `eval` here is support of more platforms. Non-standard builds of Perl are a total straw man argument.

**Comment by okurz** [comment]
@kraih you can "dismiss review" e.g. on the three dots menu in the review+check summary box at the bottom. There for example I could dismiss both your approval as well as the "change requested" by Martchus.
